File: 771.jewels-and-stones.java
/*
 * @lc app=leetcode id=771 lang=java
 *
 * [771] Jewels and Stones
 */

// @lc code=start

import java.util.HashSet;
import java.util.Set;

class Solution {
    public int numJewelsInStones(String jewels, String stones) {
        Set<Character> set = new HashSet<>();
        for (char ch : jewels.toCharArray()) {
            set.add(ch);
        }
        int ret = 0;
        for (char ch : stones.toCharArray()) {
            if (set.contains(ch)) {
                ret += 1;
            }
        }
        return ret;
    }
}
// @lc code=end

File: 771.jewels-and-stones.cpp
/*
 * @lc app=leetcode id=771 lang=cpp
 *
 * [771] Jewels and Stones
 */

// @lc code=start
#include <algorithm>
#include <cctype>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    int numJewelsInStones(string jewels, string stones) {
        vector<bool> umap(52, false);
        auto to_idx = [](char ch) -> unsigned {
            if (isupper(ch)) return ch - 'A';
            if (islower(ch)) return ch - 'a' + 26;
            return {};
        };
        for (const auto &ch : jewels) umap[to_idx(ch)] = true;
        return count_if(stones.cbegin(), stones.cend(), [&](char ch) {
            return umap[to_idx(ch)];
        });
    }
};
// @lc code=end
